### PR TITLE
Cross-dataset: gc=1.5/gc=2.0 — higher gradient clipping for DrivAerML

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AirfRANS achieved its best result at gc=1.0, which was a significant win over the no-clipping baseline. TandemFoil also benefits from gc=1.0. The question is whether DrivAerML (currently using no gradient clipping in the golden baseline) would benefit from more aggressive clipping — and specifically whether gc=1.5 or gc=2.0 (higher than the AirfRANS winner) is optimal for the larger, more complex 3D automotive case. Higher gradient clipping values may provide better regularization for the noisier, higher-dimensional DrivAerML gradients while still allowing larger gradient steps than gc=1.0.

## Instructions

Run 8 GPU jobs: 4 for gc=1.5, 4 for gc=2.0. Use `--wandb-group gc-high-sweep` so all runs are grouped.

**DrivAerML gc=1.5 runs (GPUs 0-3):**

```bash
# GPU 0
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml \
  --num-layers 4 --hidden-dim 512 --num-heads 8 \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 1.5 \
  --no-use-ema --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group gc-high-sweep \
  --wandb-name drivaer-4L-512d-gc1.5-seed0 \
  --seed 0 2>&1 | tee /tmp/casca-drivaer-gc1.5-gpu0.log &

# GPU 1
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset drivaerml \
  --num-layers 4 --hidden-dim 512 --num-heads 8 \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 1.5 \
  --no-use-ema --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group gc-high-sweep \
  --wandb-name drivaer-4L-512d-gc1.5-seed1 \
  --seed 1 2>&1 | tee /tmp/casca-drivaer-gc1.5-gpu1.log &

# GPU 2
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset drivaerml \
  --num-layers 4 --hidden-dim 512 --num-heads 8 \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 1.5 \
  --no-use-ema --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group gc-high-sweep \
  --wandb-name drivaer-4L-512d-gc1.5-seed2 \
  --seed 2 2>&1 | tee /tmp/casca-drivaer-gc1.5-gpu2.log &

# GPU 3
cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 python train.py \
  --dataset drivaerml \
  --num-layers 4 --hidden-dim 512 --num-heads 8 \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 1.5 \
  --no-use-ema --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group gc-high-sweep \
  --wandb-name drivaer-4L-512d-gc1.5-seed3 \
  --seed 3 2>&1 | tee /tmp/casca-drivaer-gc1.5-gpu3.log &
```

**DrivAerML gc=2.0 runs (GPUs 4-7):**

```bash
# GPU 4
cd target/icml2026 && CUDA_VISIBLE_DEVICES=4 python train.py \
  --dataset drivaerml \
  --num-layers 4 --hidden-dim 512 --num-heads 8 \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 2.0 \
  --no-use-ema --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group gc-high-sweep \
  --wandb-name drivaer-4L-512d-gc2.0-seed0 \
  --seed 0 2>&1 | tee /tmp/casca-drivaer-gc2.0-gpu4.log &

# GPU 5
cd target/icml2026 && CUDA_VISIBLE_DEVICES=5 python train.py \
  --dataset drivaerml \
  --num-layers 4 --hidden-dim 512 --num-heads 8 \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 2.0 \
  --no-use-ema --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group gc-high-sweep \
  --wandb-name drivaer-4L-512d-gc2.0-seed1 \
  --seed 1 2>&1 | tee /tmp/casca-drivaer-gc2.0-gpu5.log &

# GPU 6
cd target/icml2026 && CUDA_VISIBLE_DEVICES=6 python train.py \
  --dataset drivaerml \
  --num-layers 4 --hidden-dim 512 --num-heads 8 \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 2.0 \
  --no-use-ema --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group gc-high-sweep \
  --wandb-name drivaer-4L-512d-gc2.0-seed2 \
  --seed 2 2>&1 | tee /tmp/casca-drivaer-gc2.0-gpu6.log &

# GPU 7
cd target/icml2026 && CUDA_VISIBLE_DEVICES=7 python train.py \
  --dataset drivaerml \
  --num-layers 4 --hidden-dim 512 --num-heads 8 \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 2.0 \
  --no-use-ema --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group gc-high-sweep \
  --wandb-name drivaer-4L-512d-gc2.0-seed3 \
  --seed 3 2>&1 | tee /tmp/casca-drivaer-gc2.0-gpu7.log &
```

Wait for all 8 runs to finish, then report the best `val_primary/surface_rel_l2_pct` for gc=1.5 (best across seeds 0-3) and gc=2.0 (best across seeds 4-7).

## Baseline

- **DrivAerML best**: `val_primary/surface_rel_l2_pct = 4.619%` (epoch 256, PR #2691)
  - Config: 4L/512d/8H, AdamW lr=5e-4, T_max=30, **no gradient clipping**, no WD
  - Reproduce: `cd target/icml2026 && python train.py --dataset drivaerml --num-layers 4 --hidden-dim 512 --num-heads 8 --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`
- **AirfRANS best**: `val_primary/surface_mse = 0.001236` (epoch 349, PR #2828); 2L/256d/4H, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5
- **TandemFoil best**: `val_primary/surface_pressure_mae = 30.10` (epoch 157, PR #2810); 3L/192d/3H, Lion lr=1.25e-4, gc=1.0, WD=1e-2, T_max=10

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT casca:

## Results

### Summary

**gc=1.5 best: 6.066%** (seed 2, ep211) | **gc=2.0 best: 8.834%** (seed 2, ep96) | **Baseline (no clipping): 4.619%** at ep256

Both gradient clipping values are **significantly worse than the 4.619% baseline**. All 8 runs eventually diverged. The hypothesis that higher gradient clipping would help DrivAerML training is clearly rejected.

### Detailed Results

#### gc=1.5 (GPUs 0-3)

| Seed | Best val_primary/surface_rel_l2_pct | Best Epoch | Diverged At | W&B Run ID |
|------|--------------------------------------|------------|-------------|------------|
| 0    | 22.137%                              | ep22       | ep28        | f4tsy4gt   |
| 1    | 8.312%                               | ep95       | ep116       | do5kcd8f   |
| **2**| **6.066%**                           | **ep211**  | ep250       | fp4i7m0o   |
| 3    | 6.226%                               | ep195      | ep217       | 8mbrn6yv   |

**gc=1.5 best across seeds: 6.066%** (seed 2)

#### gc=2.0 (GPUs 4-7)

| Seed | Best val_primary/surface_rel_l2_pct | Best Epoch | Diverged At | W&B Run ID |
|------|--------------------------------------|------------|-------------|------------|
| 0    | 14.847%                              | ep77       | ep36        | zwc0lqxs   |
| 1    | 22.582%                              | ep22       | ep44        | strgi36m   |
| **2**| **8.834%**                           | **ep96**   | ep116 (NaN) | k5r6kf5n   |
| 3    | 8.861%                               | ep91       | ep103       | vgr2j7zm   |

**gc=2.0 best across seeds: 8.834%** (seed 2)

### Comparison to Baseline

| Config        | Best surface_rel_l2_pct | Best Epoch | Survival Rate |
|---------------|-------------------------|------------|---------------|
| No clipping   | **4.619%**              | ep256      | —             |
| gc=1.5        | 6.066% (+31.3%)         | ep211      | 1/4 (to ep250)|
| gc=2.0        | 8.834% (+91.2%)         | ep96       | 0/4           |

### Divergence Analysis

**All 8 runs diverged.** The divergence pattern is consistent:
1. Healthy convergence with low clip events (0-5 per epoch)
2. Instability develops: clip events begin escalating (14→33→75→108)
3. Grad norms spike (1.5→50→180→350+) over 2-4 epochs
4. Run reaches 394/394 clip events and grad norms hit Infinity
5. Val metric plateaus at a diverged value (60-88%) and never recovers

**gc=2.0 diverged earlier and more severely than gc=1.5:**
- gc=2.0 median divergence epoch: ~80 (range: ep36-116)
- gc=1.5 median divergence epoch: ~167 (range: ep28-250)

**Seed 2 was uniquely resilient for gc=1.5**: It survived two instability episodes (ep161-195 and ep233-248) where clip events spiked to 55-65 but recovered, before finally diverging at ep249-250 with clip events 242→387. This suggests that gradient clipping at 1.5 is right at the edge of stability for this architecture.

### GPU 2 (gc=1.5, seed 2) Convergence Trajectory
ep15: ~24% → ep45: ~12% → ep75: ~9.9% → ep105: ~8.2% → ep139: 6.66% → ep149: 6.555% → ep156: 6.295% → ep211: 6.066% → [instability] → ep225: 6.084% → [diverged ep250]

### What Happened

Gradient clipping introduces a fundamental instability on DrivAerML with this 4L/512d architecture. The clipping threshold prevents gradient norms from being expressed, but the underlying optimizer state accumulates instability over time. With CosineAnnealingWarmRestarts (T_max=30), each LR restart amplifies latent instability until it crosses a critical threshold and the run diverges catastrophically within 2-4 epochs.

gc=2.0 is too loose — it provides almost no effective clipping (the baseline without clipping already has moderate grad norms) while introducing numerical instability. gc=1.5 provides more constraint but still cannot prevent the eventual divergence.

The baseline (no clipping) at 4.619% is strictly better, suggesting that DrivAerML training with this architecture does not benefit from gradient clipping at these thresholds.

### Peak Memory
~5,985 MiB per GPU (consistent across all 8 runs)

### Commands Used
```bash
# gc=1.5 runs (GPUs 0-3)
cd /workspace/senpai/target/icml2026 && CUDA_VISIBLE_DEVICES={0,1,2,3} python train.py \
  --dataset drivaerml \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 1.5 \
  --no-use-ema --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group gc-high-sweep \
  --wandb-name "casca/drivaer-4L-512d-gc1.5-seed{0-3}-v2" \
  --agent casca \
  --seed {0-3} --epochs 999

# gc=2.0 runs (GPUs 4-7) — same as above with --grad-clip 2.0
```

### W&B
- Group: `gc-high-sweep`
- All 8 runs visible in group

### Suggested Follow-ups
1. **Lower gradient clipping values** (gc=0.5, gc=1.0) may be worth investigating — tighter clipping might prevent the divergence pattern while still providing regularization benefit.
2. **Gradient clipping + lower learning rate** — the instability is triggered by LR restarts, so a lower base LR (e.g., 2e-4) combined with gc=1.5 might avoid the critical threshold.
3. **Adaptive gradient clipping** — methods that scale the clip threshold by parameter norm (like AGC from NFNet) may be better suited to this architecture than fixed-threshold clipping.
4. **Different LR schedule** — CosineAnnealingWarmRestarts amplifies instability at restarts. A monotonic schedule (cosine decay without restarts) might pair better with gradient clipping.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2881 because ADVISOR: Closing — clear dead end. gc=1.5 best: 6.066% (+31.3% vs 4.619% baseline), gc=2.0 best: 8.834% (+91.2% vs baseline). All 8 runs diverged. CosineAnnealingWarmRestarts LR restarts amplify instability when clipping is active. Thank you for the thorough analysis — your AGC (Adaptive Gradient Clipping) suggestion is well-motivated and will be assigned as a follow-up..
